Zend/zend_hash.c

```diff
@@ -480,6 +480,31 @@ ZEND_API void ZEND_FASTCALL _zend_hash_iterators_update(HashTable *ht, HashPosit
 	}
 }
 
+/* The macro casts strings to be compared to zend_long arrays truncating the length to multiple of
+ * zend_long size. This is acceptable for hash string comparisons as the ignored tailing bytes are already
+ * implicitly checked by hash value matching.
+ * Hash value function ensures (see zend_inline_hash_func() in zend_string.h):
+ * - strings with lengths < SIZEOF_ZEND_LONG will generate unique hash values
+ * - strings with identical length differing only in last (SIZEOF_ZEND_LONG-1) bytes will always generate
+ *   different hash values.
+ */
+#define ZEND_STR2LONG_MEMEQ(s1, s2 ,len ) \
+		zend_ulong_memeq((const zend_ulong *)(s1), (const zend_ulong *)(s2), (len) >> (SIZEOF_ZEND_LONG_LOG2))
+
+/* returns 1 if the zend_log arrays are equal, 0 otherwise; empty arrays are considered equal. */
+static zend_always_inline zend_bool zend_ulong_memeq(const zend_ulong *arr1, const zend_ulong *arr2, size_t len)
+{
+	while (len != 0) {
+		if ( *arr1 != *arr2 ) {
+			return 0;
+		}
+		arr1++;
+		arr2++;
+		len --;
+	}
+	return 1; /* equal arrays */
+}
```

Comment:
This function works better than libc memcpy for small strings with multiple of zend_ulong size which is the usual case for hash tables; 
The advantage stays in assuming there are no trailing bytes outside of 32/64bit chunks, so it does not have to deal with them; for short strings (which is the common case for hash tables) dealing with trailing bytes takes much more than the 2-3 loops of integer chunks; SIMD benefit will be visible just in case of long strings which are rare our case.
In the concrete case of hash tables lookups I don't know how could the compiler optimize memcmp() as long as the string sizes are unknown at compile time.

I did some measurements with VTune Amplifier and I saw zend_hash_find_bucket() CPU cycles reduced at half (from 6% to 3%) in WordPress.